src/lcdlib/lcdlib_common_oled.c

```diff
@@ -0,0 +1,531 @@
+        free(fb_buffer);
+        fb_buffer = NULL;
+        close(fb_fd);
+        fb_fd = -1;
+        return LCD_ERR_READ;
+    }
+
+    int offset = y * OLED_WIDTH_PX / 2 + x / 2; // Calculate the position in the buffer
+
+    // Clear Line to be written:
+    lcdlib_clear_rows(y, LINE_HEIGHT_PX);
+
+    // 128 pixels in each line, each pixel is 4 bits (1 nibble)
+    for (int k = 0; k < str_len && *str; k++)
+    {
+        unsigned char glyph[CHARACTER_WIDTH_PX]; // 6x8 glyphs
+        load_glyph(*str, glyph);                 // Load glyph for character
+
+        for (int i = 0; i < CHARACTER_HEIGHT_PX; i++)
+        { // Iterate over rows of the glyph
+            for (int j = 0; j < CHARACTER_WIDTH_PX; j++)
+            {                                             // Iterate over columns of the glyph
+                int byte_index = offset + i * OLED_WIDTH_BYTES + j / 2; // Correct byte in the buffer
+                int pixel_value = (glyph[j] >> i) & 1 ? 0xF : 0x0;
+
+                if (j % 2 == 0)
+                {
+                    // Even column, modify the high nibble (4 bits)
+                    fb_buffer[byte_index] = (fb_buffer[byte_index] & 0x0F) | (pixel_value << 4);
+                }
+                else
+                {
+                    // Odd column, modify the low nibble (4 bits)
+                    fb_buffer[byte_index] = (fb_buffer[byte_index] & 0xF0) | pixel_value;
+                }
+            }
+        }
+
+        offset += CHARACTER_WIDTH_PX / 2; // Move to the next character position (6 pixels, 3 bytes)
+        str++;
+    }
+
+    // Write the buffer to the framebuffer device
+    lseek(fb_fd, 0, SEEK_SET);
+    ssize_t bytes_written = write(fb_fd, fb_buffer, OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES);
+    if (bytes_written != OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES)
+    {
+        print_err("Error: failed to write framebuffer content\n");
+        return LCD_ERR_WRITE;
+    }
+
+    return 0;
+}
+
+int lcdlib_open_dev(void)
+{
+    fb_fd = open("/dev/fb0", O_RDWR);
+    if (fb_fd < 0)
+    {
+        print_err("Error: cannot open framebuffer device\n");
+        return -1;
+    }
+
+    // Allocate buffer to hold one frame
+    fb_buffer = (unsigned char *)malloc(OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES);
+    if (!fb_buffer)
+    {
+        print_err("Error: cannot allocate framebuffer memory\n");
+        close(fb_fd);
+        return -1;
+    }
+
+    // Read the current framebuffer content into the buffer
+    lseek(fb_fd, 0, SEEK_SET);
+    ssize_t bytes_read = read(fb_fd, fb_buffer, OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES);
+    if (bytes_read != OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES)
+    {
+        print_err("Error: failed to read framebuffer content\n");
+        free(fb_buffer);
+        fb_buffer = NULL;
+        close(fb_fd);
+        fb_fd = -1;
+        return LCD_ERR_READ;
+    }
+
+    return 0;
+}
+int lcdlib_close_dev(void)
```

Comment:
This API signature needs change to extend beyond single buffer.
something like int lcdlib_close_dev(char *deviceName) or int lcdlib_close_dev(int deviceId)